packages/frontend/src/components/ChooseConnectionSubstep/index.tsx

```diff
@@ -61,37 +70,37 @@ function ChooseConnectionSubstep(
     onSubmit,
     onChange,
     application,
+    selectedActionOrTrigger,
   } = props
   const { connection, appKey } = step
   const formatMessage = useFormatMessage()
-  const editorContext = React.useContext(EditorContext)
-  const [showAddConnectionDialog, setShowAddConnectionDialog] =
-    React.useState(false)
+  const editorContext = useContext(EditorContext)
+  const [showAddConnectionDialog, setShowAddConnectionDialog] = useState(false)
   const { data, loading, refetch } = useQuery(GET_APP_CONNECTIONS, {
     variables: { key: appKey },
   })
-  // TODO: show detailed error when connection test/verification fails
-  const [
-    testConnection,
-    { loading: testResultLoading, refetch: retestConnection },
-  ] = useLazyQuery(TEST_CONNECTION, {
+  const supportsWebhookRegistration =
+    (selectedActionOrTrigger as ITrigger)?.supportsWebhookRegistration || false
+
+  const {
+    loading: testResultLoading,
+    refetch: retestConnection,
+    data: testConnectionData,
+  } = useQuery<{ testConnection: ITestConnectionResult }>(TEST_CONNECTION, {
     variables: {
-      id: connection?.id,
+      connectionId: connection?.id,
+      stepId: supportsWebhookRegistration ? step.id : undefined,
     },
+    skip: !connection?.id,
   })
 
-  React.useEffect(() => {
-    if (connection?.id) {
-      testConnection({
-        variables: {
-          id: connection.id,
-        },
-      })
+  useEffect(() => {
+    if (step.connection?.id) {
+      retestConnection()
```

Comment:
Wait, can we omit the variables argument? I thought apollo would use the value of the previous invocation's argument if it was omitted, but we're triggering the refetch if connection ID changed.

Reply:
the useQuery hook should reinstantiate with the new connectionId when it changes. I've tested it out on the frontend. Does it not work for u?

Reply:
Yes it works but I was wondering _why_ it worked. I guess Apollo [doc](https://www.apollographql.com/docs/react/data/queries/#providing-new-variables-to-refetch) is worded misleadingly; it says:

> `refetch` uses each omitted variable's original value.

but that doesn't seem apply at all; it always uses the currently selected connection ID even if I pass in something like:
```
{
  variables: {
    stepId: step.id,
  },
}
```

I haven't read apollo source to confirm.... but I go file an issue in apollo first. wth is this.